rts/Rml/Backends/RmlUi_Backend.cpp

```diff
@@ -0,0 +1,495 @@
+  Lifetime governed by the calls to Backend::Initialize() and
+  Backend::Shutdown().
+ */
+struct BackendData {
+	RmlSystemInterface system_interface;
+#ifndef HEADLESS
+	RenderInterface_GL3_Recoil render_interface;
+#else
+	RenderInterface_Headless render_interface;
+#endif
+	VFSFileInterface file_interface;
+
+	SDL_Window* window = nullptr;
+	SDL_GLContext gl_context = nullptr;
+	std::vector<Rml::Context*> contexts;
+	InputHandler::SignalType::connection_type inputCon;
+	CRmlInputReceiver inputReceiver;
+
+	// make atomic_bool?
+	bool initialized = false;
+	bool debuggerAttached = false;
+	int winX = 1;
+	int winY = 1;
+
+	lua_State* ls = nullptr;
+	Rml::SolLua::SolLuaPlugin* luaPlugin = nullptr;
+
+	CtxMutex contextMutex;
+	Rml::UniquePtr<PassThroughPlugin> plugin;
+    Rml::UniquePtr<Rml::ElementInstancerGeneric<RmlGui::ElementLuaTexture>> element_lua_texture_instancer;
+};
+
+static Rml::UniquePtr<BackendData> data;
+
+bool RmlInitialized()
+{
+	return data && data->initialized;
+}
+
+bool RmlGui::Initialize(SDL_Window* target_window, SDL_GLContext target_glcontext, int winX,
+                        int winY)
+{
+	data = Rml::MakeUnique<BackendData>();
+
+	if (!data->render_interface) {
+		data.reset();
+		fprintf(stderr, "Could not initialize OpenGL3 render interface.");
+		return false;
+	}
+
+	data->window = target_window;
+	data->gl_context = target_glcontext;
+
+	Rml::SetFileInterface(&data->file_interface);
+	Rml::SetSystemInterface(RmlGui::GetSystemInterface());
+	Rml::SetRenderInterface(RmlGui::GetRenderInterface());
+
+	data->render_interface.SetViewport(winX, winY);
+	data->winX = winX;
+	data->winY = winY;
+
+	Rml::Initialise();
+
+	Rml::LoadFontFace("fonts/FreeSansBold.otf", true);
+	data->inputCon = input.AddHandler(&RmlGui::ProcessEvent);
+	data->initialized = true;
+
+	data->element_lua_texture_instancer = Rml::MakeUnique<Rml::ElementInstancerGeneric<ElementLuaTexture>>();
+	Rml::Factory::RegisterElementInstancer("texture", data->element_lua_texture_instancer.get());
+
+	data->plugin = Rml::MakeUnique<PassThroughPlugin>(OnContextCreate, OnContextDestroy);
+	Rml::RegisterPlugin(data->plugin.get());
+
+	return true;
+}
+
+bool RmlGui::InitializeLua(lua_State* lua_state)
+{
+	if (!RmlInitialized()) {
+		return false;
+	}
+	sol::state_view lua(lua_state);
+	data->ls = lua_state;
+	data->luaPlugin = Rml::SolLua::Initialise(&lua, "rmlDocumentId");
+	data->system_interface.SetTranslationTable(&data->luaPlugin->translationTable);
+	return true;
+}
+
+bool RmlGui::RemoveLua()
+{
+	if (!RmlInitialized() || !data->ls) {
+		return false;
+	}
+	data->luaPlugin->RemoveLuaItems();
+	Update();
+	Rml::UnregisterPlugin(data->luaPlugin);
+	data->system_interface.SetTranslationTable(nullptr);
+	data->luaPlugin = nullptr;
+	data->ls = nullptr;
+
+	return true;
+}
+
+void RmlGui::Shutdown()
+{
+	if (!RmlInitialized()) {
+		return;
+	}
+
+	Rml::UnregisterPlugin(data->plugin.get());
+	data.reset();
+
+	// removes all contexts
+	Rml::Shutdown();
```

Comment:
According to https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/integrating.html#the-system-interface (emphasis mine):

> The function should return the time (in seconds) since the start of the application. Install your system interface by calling Rml::SetSystemInterface() with a pointer to the interface. **Note that you must keep the system interface alive until after the call to Rml::Shutdown() and destroy it afterwards.** RmlUi won’t release your interfaces.

The system interface (and render interface) are, I believe, destroyed in `data.reset()`, which I believe is at odds with the ordering here. You can also see in the sample at https://github.com/mikke89/RmlUi/blob/master/Samples/invaders/src/main.cpp#L130-L133 that they call `Rml::Shutdown();` before `Backend::Shutdown();`.

Reply:
You are correct, the ordering needs to be corrected

Reply:
This has been resolved :+1: .